initrd/bin/oem-factory-reset

```diff
@@ -437,343 +826,411 @@ usb_security_token_capabilities_check
+        fi
     fi
 fi
 
 assert_signable
 
 # Action time...
 
+# clear gpg-agent cache so that next gpg calls doesn't have past keyring in memory
+killall gpg-agent >/dev/null 2>&1 || true
+# clear local keyring
+rm -rf /.gnupg/* >/dev/null 2>&1 || true
+
 # detect and set /boot device
 echo -e "\nDetecting and setting boot device...\n"
-if ! detect_boot_device ; then
-  SKIP_BOOT="y"
+if ! detect_boot_device; then
+    SKIP_BOOT="y"
 else
-  echo -e "Boot device set to $CONFIG_BOOT_DEV\n"
+    echo -e "Boot device set to $CONFIG_BOOT_DEV\n"
 fi
 
 # update configs
 if [[ "$SKIP_BOOT" == "n" ]]; then
-  replace_config /etc/config.user "CONFIG_BOOT_DEV" "$CONFIG_BOOT_DEV"
-  combine_configs
+    replace_config /etc/config.user "CONFIG_BOOT_DEV" "$CONFIG_BOOT_DEV"
+    combine_configs
 fi
 
-
 if [ -n "$luks_new_Disk_Recovery_Key_desired" -a -n "$luks_new_Disk_Recovery_Key_passphrase_desired" ]; then
-  #Reencryption of disk, disk recovery key and Disk Recovery Key passphrase change is requested
-  luks_reencrypt
-  luks_change_passphrase
+    #Reencryption of disk, disk recovery key and Disk Recovery Key passphrase change is requested
+    luks_reencrypt
+    luks_change_passphrase
 elif [ -n "$luks_new_Disk_Recovery_Key_desired" -a -z "$luks_new_Disk_Recovery_Key_passphrase_desired" ]; then
-  #Reencryption of disk was requested but not passphrase change
-  luks_reencrypt
+    #Reencryption of disk was requested but not passphrase change
+    luks_reencrypt
 elif [ -z "$luks_new_Disk_Recovery_Key_desired" -a -n "$luks_new_Disk_Recovery_Key_passphrase_desired" ]; then
-  #Passphrase change is requested without disk reencryption
-  luks_change_passphrase
+    #Passphrase change is requested without disk reencryption
+    luks_change_passphrase
 fi
 
 ## reset TPM and set password
 if [ "$CONFIG_TPM" = "y" ]; then
-  echo -e "\nResetting TPM...\n"
-  tpmr reset "$TPM_PASS" >/dev/null 2>/tmp/error
+    echo -e "\nResetting TPM...\n"
+    tpmr reset "$TPM_PASS" >/dev/null 2>/tmp/error
 fi
 if [ $? -ne 0 ]; then
-	ERROR=$(tail -n 1 /tmp/error | fold -s)
-	whiptail_error_die "Error resetting TPM:\n\n${ERROR}"
+    ERROR=$(tail -n 1 /tmp/error | fold -s)
+    whiptail_error_die "Error resetting TPM:\n\n${ERROR}"
 fi
 
 # clear local keyring
 rm /.gnupg/*.gpg 2>/dev/null
 rm /.gnupg/*.kbx 2>/dev/null
+# initialize gpg wth empty keyring
 gpg --list-keys >/dev/null 2>&1
 
-## reset the GPG Key
-echo -e "\nResetting GPG Key...\n(this will take around 3 minutes...)\n"
-gpg_key_reset
+#Generate keys in memory and copy to smartcard
+if [ "$GPG_GEN_KEY_IN_MEMORY" == "1" ]; then
+    if [ "$GPG_ALGO" == "RSA" ]; then
+        # Generate GPG master key
+        generate_inmemory_RSA_master_and_subkeys
+        wipe_thumb_drive_and_copy_gpg_key_material
+        #TODO seperate setting config. Recovery shell under qemu without reboot doesn't prompt for PIN
+        set_user_config "CONFIG_HAVE_GPG_KEY_BACKUP" "y"
+        keytocard_subkeys_to_smartcard
```

Comment:
Everything from `wipe_thumb_drive_and_...` to `keytocard_...` should not depend on the GPG_ALGO AFAICT, move that out of the if-else.

Is this why recovery isn't prompting for gpg_auth for you - were you generating a p256 key by chance?

Reply:
@JonathonHall-Purism That was it :) thanks for your eyes. Fixed under [c4a94b3](https://github.com/osresearch/heads/pull/1515/commits/c4a94b34ab9a69efa8298b0fd4e406cfc9831b18) which can be tested from reboot->recovery->gpg_auth as of current qemu board configs. Also note that this commit permits to generate in-memory keygen+key backup to be usable without using any USB Security dongle.

@JonathonHall-Purism Thnaks for having finding logic error here, my following steps will be made much more easier without relaying on qubes->usb->USB Security dongle to use this.